inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -4113,6 +4113,9 @@ static final class BeanDefinitionProducer {
 
         BeanDefinitionProducer(@NonNull BeanDefinitionReference reference) {
             this.reference = reference;
+            if (reference instanceof BeanDefinition<?>) {
```

Comment:
you should also add an `instanceof AbstractInitializableBeanDefinition` check, it will be faster and should be hit most of the time.